### PR TITLE
Epoch frontier state sync

### PR DESCRIPTION
**Pull request by Kouprin**
Fixes https://github.com/nearprotocol/nearcore/issues/2021.

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -1311,13 +1318,20 @@ impl Chain {
...
         if shard_id as usize >= sync_block.chunks.len() {
             return Err(ErrorKind::Other(
                 "get_syncing_state_root fail: shard_id out of bounds".into(),
             )
             .into());
         }
         let sync_prev_block = self.get_block(&sync_block_header.prev_hash)?;
+        if sync_block_epoch_id == sync_prev_block.header.inner_lite.epoch_id {
+            return Err(ErrorKind::Other(
+                "Invalid request: sync_hash is not the first hash of the epoch".into(),
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs]: looks like some repetition here

File: chain/client/src/sync.rs [hunk 2]
```diff
@@ -672,6 +672,21 @@ impl StateSync {
         Ok((update_sync_status, all_done))
     }
 
+    fn get_epoch_start_sync_hash(
+        &self,
+        chain: &mut Chain,
+        sync_hash: CryptoHash,
```
Comment (bowenwang1996) [inline chain/client/src/sync.rs]: why not pass as reference to avoid unnecessary copying?

File: chain/client/src/sync.rs [hunk 3]
```diff
@@ -672,6 +672,21 @@ impl StateSync {
...
+    ) -> Result<CryptoHash, near_chain::Error> {
+        let mut hash = sync_hash;
+        loop {
+            let header = chain.get_block_header(&hash)?;
+            if header.inner_lite.next_epoch_id.0 == header.prev_hash {
```
Comment (bowenwang1996) [inline chain/client/src/sync.rs]: In general we should avoid compare epoch id with block hash. Use methods exposed in runtime adapter instead.

File: chain/chain/src/chain.rs (hunk 1 above)
Comment (Kouprin) [reply chain/chain/src/chain.rs]: ++

File: chain/client/src/sync.rs (hunk 3 above)
Comment (Kouprin) [reply chain/client/src/sync.rs]: @bowenwang1996 could you please point which exact methods are better to use instead?

File: chain/client/src/sync.rs (hunk 2 above)
Comment (Kouprin) [reply chain/client/src/sync.rs]: ++

**Comment by Kouprin** [comment]
> I don't quite get the point of changing sync_hash to always be the hash of the first block of an epoch.

This question is related to the issue, not to the PR. Could you please go to https://github.com/nearprotocol/nearcore/issues/2021 and discuss there?

**Comment by Kouprin** [comment]
@SkidanovAlex @bowenwang1996 
I'm afraid of some unexpected circumstances of changing sync_hash to the first hash of the epoch. For example, if horizons are chosen arbitrary, it's possible to get stuck in infinite loop of state syncing. Personally, I prefer to replace sync_hash with next_epoch_id (https://github.com/nearprotocol/nearcore/pull/2068) but it touches many pieces of code while refactoring.

File: chain/client/src/sync.rs (hunk 2 above)
Comment (Kouprin) [reply chain/client/src/sync.rs]: @bowenwang1996 I found that I don't understand what are you talking about. What are we saving at this point?

**Comment by Kouprin** [comment]
ErrorKind updated. Waiting for more explanations from @bowenwang1996

File: chain/client/src/sync.rs (hunk 3 above)
Comment (SkidanovAlex) [reply chain/client/src/sync.rs]: Given that you are already in a loop, just comparing epoch id to the prev epoch id should be sufficient?

**Comment by SkidanovAlex** [comment]
The point re: infinite state sync is good.
We should use a combination of the horizon and the epoch id (only trigger state sync iff the current head is more than horizon back, and has a different epoch ID that the header head)

File: chain/client/src/sync.rs (hunk 3 above)
Comment (Kouprin) [reply chain/client/src/sync.rs]: ++

File: chain/client/src/sync.rs (hunk 2 above)
Comment (Kouprin) [reply chain/client/src/sync.rs]: I've changed to `sync_hash: &CryptoHash` as you suggested, however, it doesn't make any sense to me.

**Comment by bowenwang1996** [comment]
If we only sync to epoch start we probably need to change how block sync works. With this change I fear that block sync wouldn't actually kick in, which means then we just rely on orphan sync after state sync is done. Also make sure that we don't have infinite state sync :)

**Comment by SkidanovAlex** [comment]
We also  need to change the logic that decides whether the state sync is needed, to only switch to state sync iff:
a) The block head is outside the horizon from the header head (this part already exists), *and*
b) The block head and the header head are in different epochs.

**Pull request merged**